polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1477,6 +1479,148 @@ class RaiseSetNDRange
+    auto enclosingFunc = op->getParentOfType<LLVM::LLVMFuncOp>();
+    if (!enclosingFunc || !getHandler(enclosingFunc))
+      return failure();
+
+    // It must be a "kernel" annotation.
+    FailureOr<StringRef> failureOrAnnotationStr = getAnnotation(op);
+    if (failed(failureOrAnnotationStr) || *failureOrAnnotationStr != "kernel")
+      return failure();
+
+    // There should be exactly one store to the marked pointer.
+    Value annotatedPtr = op.getVal();
+    auto [store, lambdaObj] =
+        getUniqueAssignment(annotatedPtr, /*allowMemcpy=*/false);
+    if (!store)
+      return failure();
+
+    auto alloca = dyn_cast_or_null<LLVM::AllocaOp>(lambdaObj.getDefiningOp());
+    if (!alloca || !alloca.getElemType().has_value())
+      return failure();
+
+    auto lambdaClassTy =
+        dyn_cast<LLVM::LLVMStructType>(alloca.getElemType().value());
+    if (!lambdaClassTy)
+      return failure();
+
+    auto captureTypes = lambdaClassTy.getBody();
+
+    // Look for unique assignments that represent the capturing of kernel args.
+    // Assuming the no-op `getelementptr %lambdaObj[0, 0]` was folded, the
+    // first capture is actually a store to the pointer.
+    if (auto [captureOp, capturedVal] = getUniqueAssignment(lambdaObj);
+        captureOp) {
+      capturedVal = tryToBroaden(capturedVal, captureTypes[0]);
+      rewriter.setInsertionPointAfter(captureOp);
+      rewriter.create<sycl::SYCLHostSetCaptured>(captureOp->getLoc(), lambdaObj,
+                                                 rewriter.getI64IntegerAttr(0),
+                                                 capturedVal);
+    }
+
+    // All other captures are unique assignments to GEPs with two constant
+    // indices `[0, <capture #>]` to the lambda object.
+    for (auto *user : lambdaObj.getUsers()) {
+      auto gep = dyn_cast<LLVM::GEPOp>(user);
+      if (!gep || !gep.getElemType().has_value() ||
+          gep.getElemType().value() != lambdaClassTy ||
+          !gep.getDynamicIndices().empty())
+        continue;
+
+      auto indices = gep.getRawConstantIndices();
+      if (indices.size() != 2 || indices[0] != 0)
+        continue;
+      unsigned captureIdx = indices[1];
+
+      auto [captureOp, capturedVal] = getUniqueAssignment(gep);
+      if (!captureOp)
+        continue;
+
+      capturedVal = tryToBroaden(capturedVal, captureTypes[captureIdx]);
+      rewriter.setInsertionPointAfter(captureOp);
+      rewriter.create<sycl::SYCLHostSetCaptured>(
+          captureOp->getLoc(), lambdaObj,
+          rewriter.getI64IntegerAttr(captureIdx), capturedVal);
+    }
+
+    // Finally erase the annotation op.
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+private:
+  /// If there is exactly one store or memcpy to \p ptr, return it and the
+  /// stored value/copied-from pointer; otherwise return `nullptr` and an emtpy
+  /// value.
+  static std::tuple<Operation *, Value>
+  getUniqueAssignment(Value ptr, bool allowMemcpy = true) {
```

Comment:
`allowMemcpy` is not used.

Reply:
I removed that argument in b802138d2ac842ee1f32a142a24fe27f1a35eeba; the check is now done at the call-site.